### ISSUE TITLE
[FIX-001] Normalize built-in logits before negative-log-likelihood loss

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#200]
**Category:** Training and model selection · **Priority:** P1 · **Size:** M

**Evidence status:** Reproduced defect.

## Outcome and scope

Built-in classifiers configured with native NLL must optimize normalized log probabilities while continuing to expose raw logits through prediction. Repair the loss-input boundary in supervised execution and NNx-owned classifier step factories; do not change the `Losses` factory, model output activations, imperative custom-step ownership, or introduce the future task/engine features.

## Current behavior / evidence

**Current behavior:** `Losses.NEGATIVE_LOG_LIKELIHOOD` creates `torch.nn.NLLLoss`, but built-in nets emit unrestricted logits and `_loss_terms` passes them straight into the loss. The installed PyTorch `NLLLoss` docstring explicitly requires log probabilities. The factory test verifies the returned loss class, not the integration contract.

**Concrete reproduction:** Construct a CPU feed-forward NNModel with `input_dim=output_dim=2`, no hidden layers, and NLL configured. Zero the last layer's weights and set bias to `[2., 1.]`. Evaluate two zero input rows with labels `[0,0]`. The reported loss is **-2.0**. Correct normalized NLL is `F.nll_loss(F.log_softmax(logits, dim=1), labels) = 0.31326166`. The same unnormalized objective is backpropagated by the default training step, so it pushes the target logit upward without the competing-class normalization gradient.

## Repository map

- [`src/nnx/nn/nn_model.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/nn_model.py#L403) — `_loss_terms`, `default_train_step`, `evaluate`, and `_fwd_pass` are the three loss routes and the token/graph reshaping boundary.
- [`src/nnx/nn/enum/losses.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/enum/losses.py#L8) — `Losses.__call__` must still return native `torch.nn.NLLLoss`.
- [`src/nnx/nn/net/feed_fwd_nn.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/net/feed_fwd_nn.py#L23) — `FeedFwdNN.forward` returns its final Linear output without normalization.
- [`tests/test_losses.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_losses.py#L32) — Factory-type compatibility coverage.
- [`tests/test_pass2_n_series.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_pass2_n_series.py#L157) — Weighted/ignored evaluation denominators and sum-reduction coverage.
- [`tests/test_train_step_hook.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_train_step_hook.py#L41) — Public training fixture and default-step integration conventions.

## Implementation approach

1. Add one private loss-input adapter in `nn_model.py`. For the exact native `torch.nn.NLLLoss` type, apply `torch.nn.functional.log_softmax(logits, dim=1)` after `_fwd_pass` has reshaped supported Transformer token outputs to `(tokens, classes)`. Return other loss inputs unchanged. Do not infer semantics from an arbitrary subclass: preserve its supplied-input contract and document that boundary.
2. Use that adapter in `_loss_terms` and in both direct legacy `default_train_step` branches where `accumulation_state is None` (CUDA AMP and ordinary CPU). Evaluation already calls `_loss_terms`; avoid applying the adapter twice. Leave `_fwd_pass` and `PredictResult.logits` raw so decoding and callers are unaffected.
3. Continue calling the configured native loss object so its class weights, ignore index and reduction remain authoritative. Keep `_loss_normalization_weight` and zero-valid-target accumulation behavior intact; normalization must use the existing effective denominator, not batch count. Do not silently add support for nonscalar `reduction="none"`.
4. Add direct numeric oracles, not merely NLL-versus-NLL split comparisons: existing N-series tests can agree with themselves while optimizing the wrong objective. Test native NLL and CE from identical initial state and compare every updated parameter.

## Acceptance criteria

- [ ] The example above reports 0.31326166 within float32 tolerance in both default training and evaluation.
- [ ] With the same initial weights/data, NLL and cross-entropy produce matching loss and parameter updates for ordinary classifiers, including unequal accumulation batches.
- [ ] Native NLL class weights, `ignore_index`, and supported mean/sum reduction behavior remain correct against a direct PyTorch reference.
- [ ] Transformer token logits use the actual class axis after supervised reshaping; classification accuracy/predictions are unchanged.
- [ ] `predict().logits` remains the network's raw output, and custom loss modules retain a documented input contract.

## Validation and completion

Minimal existing-API fixture (the assertion describes the corrected result):

```python
import torch
from torch.utils.data import DataLoader, TensorDataset
from nnx import Devices, Losses, Nets, NNModel, NNModelParams, NNParams

model = NNModel(
    net_params=NNParams(input_dim=2, output_dim=2, hidden_dims=[], dropout_prob=0),
    params=NNModelParams(net=Nets.FEED_FWD, device=Devices.CPU,
                         loss=Losses.NEGATIVE_LOG_LIKELIHOOD),
)
with torch.no_grad():
    model.net.layers[-1].weight.zero_()
    model.net.layers[-1].bias.copy_(torch.tensor([2., 1.]))
x, y = torch.zeros(2, 2), torch.zeros(2, dtype=torch.long)
loss = model.evaluate(DataLoader(TensorDataset(x, y), batch_size=2)).loss
assert abs(loss - 0.31326166) < 1e-6
assert model.predict(x).logits.tolist() == [[2., 1.], [2., 1.]]
```

Proposed regressions: `test_nll_normalizes_builtin_logits`, `test_nll_default_step_matches_cross_entropy`, and `test_nll_preserves_custom_subclass_input`. Exercise uneven microbatches `[2, 2, 1]`, a short final accumulation window, class weights `[1, 7]`, ignored targets and `mean`/`sum`; reference `F.nll_loss(F.log_softmax(raw, dim=1), target, ...)`. Include token logits shaped `(B,T,V)` before flattening and a direct `TrainStepContext` without accumulation state. CPU coverage is mandatory; actual CUDA is required before declaring the AMP branch validated.

From the repository root, use `uv sync --frozen --all-extras` for the CI-equivalent environment. Existing focused coverage (files already exist):

```bash
uv run --frozen pytest tests/test_losses.py tests/test_pass2_n_series.py tests/test_train_step_hook.py
```

Add the proposed regression cases above to those existing test files, then rerun the same command. Finish with:

```bash
uv run --frozen ruff check src/ tests/ examples/ scripts/
uv run --frozen ruff format --check src/ tests/ examples/ scripts/
uv run --frozen pyright --warnings
```

No new dependency is needed for the implementation described. Preserve Python 3.10 compatibility and default-field omission in existing serialization. These commands are completion requirements for implementation; this specification review does not claim they were run or passed.

## Cascading changes and affected consumers

- **Owned classifier callers.** Trace every raw-logit loss call beyond the default loop: [KD and feature-KD](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/paradigms/distillation.py), [MoE](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/paradigms/moe.py) and [Mixup/CutMix](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/paradigms/augmentation.py) all invoke `m.loss_fn` directly. Reuse the corrected native-NLL input adapter for their hard/supervised terms, retaining each factory’s existing KL, auxiliary loss, interpolation weights, clipping and update ownership. Born-again training consumes the KD factory, so its NLL generations must inherit the same fix rather than add another transform. Keep decoded metrics and teacher/student soft logits raw.

- **Persistence and independent loss APIs.** Through [evaluate/train/checkpoint reconstruction](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/nn_model.py), corrected scalar losses must reach IDP history, BEST selection and plateau scheduling without adding a serialized flag or changing Losses enum values. A reloaded native-NLL checkpoint must evaluate with the new numerical contract while retaining its original descriptor. [lr_finder](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/lr_finder.py) accepts an arbitrary caller loss function; keep that contract and show callers how to supply log-softmax for NLL instead of secretly transforming all standalone outputs. Arbitrary custom hooks remain caller-owned.

- **Executable teaching path.** Extend [the classifier example](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/01_synthetic_classification.py) with a bounded native-NLL variant that trains, reloads BEST, and compares evaluation against explicit log-softmax/NLL. Add owned-factory integration cases to distillation/augmentation/MoE tests, including one shared initial-state CE reference per family. Test hard terms directly so KL or auxiliary terms cannot hide a wrong NLL sign. Update the NLL and KD explanations, which currently claim NLL compatibility, rather than retaining a supervised-only statement that contradicts shipped factory behavior.

### Samples and executable coverage

- [`examples/01_synthetic_classification.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/01_synthetic_classification.py): Add a small callable NLL path using two fixed-logit rows, one short training epoch and BEST reload, asserting explicit normalized loss and raw prediction logits. **Verify:** Proposed execution node: uv run --frozen pytest tests/test_examples_smoke.py -k nll_example_execution. CPU, offline, temporary working directory; call the bounded helper rather than the full training main. Retain import-without-main coverage and add the callable to explicit execution coverage; import success alone is insufficient.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`docs/concepts.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/concepts.md) | `generated/site/Concepts.md` | `generated/wiki/Concepts.md` | Document raw network logits versus native NLL log-probability adaptation in enum/loss and KD sections; distinguish NNx-owned classifier factories from arbitrary custom losses and standalone lr_finder. |
| [`docs/quickstart.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/quickstart.md) | `generated/site/Quickstart.md` | `generated/wiki/Quickstart.md` | Add a native-NLL variation and explain that prediction stays raw logits, evaluation/training normalization is internal, and historical NLL checkpoints use corrected loss semantics. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] NNx-owned KD, feature-KD, MoE, Mixup and CutMix hard/supervised native-NLL terms agree with explicit log-softmax references without changing their auxiliary losses or update ownership.
- [ ] A public native-NLL train → history save/load → BEST reconstruction → evaluate flow reports corrected losses while keeping enum/config serialization and raw prediction logits unchanged.
- [ ] An explicit user-defined train hook and standalone lr_finder loss callable receive their original inputs; the documented NLL callable recipe supplies its own log-softmax.
- [ ] The bounded classifier-example NLL variant executes in a temporary directory and verifies its reloaded BEST evaluation numerically, not merely by importing the script.
- [ ] The issue-specific changes in `docs/concepts.md`, `docs/quickstart.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** Propagation belongs to this fix and existing APIs. Related tickets coordinate shared contracts/files; they are not prerequisites for the bounded repaired path.

Additional nonblocking consumers/coordination: [FEAT-002](https://github.com/thekaveh/NNx/issues/212), [FEAT-004](https://github.com/thekaveh/NNx/issues/214).

## Delivery order and dependencies

- **Execution order:** 01 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.
- **Related / optional integration:** [FEAT-002](https://github.com/thekaveh/NNx/issues/212), [FEAT-004](https://github.com/thekaveh/NNx/issues/214). These links do not block this ticket.

**Dependency rationale.** Independent repair of current behavior. No dependency; FEAT-002/004 may later reuse the corrected loss-input contract.

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved.

**Root cause:** `Losses.NEGATIVE_LOG_LIKELIHOOD` builds `torch.nn.NLLLoss` (which requires log-probabilities), but built-in nets emit raw logits and every loss route (`_loss_terms`, both legacy `default_train_step` branches, and the owned KD/feature-KD/MoE/Mixup/CutMix factories) passed them straight through — the reported loss could go negative and the gradient lacked the competing-class normalization term.

**What changed**
- `_is_native_nll` (extracted from the existing predicate) + private `_loss_input(loss_fn, logits)` adapter: `log_softmax(logits, dim=1)` for the exact native `NLLLoss` after `_fwd_pass` has flattened token logits to `(rows, classes)`; raw output for every other loss. Used in `_loss_terms` (evaluate + accumulation path) and both `accumulation_state is None` branches. Class weights, `ignore_index`, reduction and the effective-denominator normalization stay authoritative; `predict().logits` stays raw.
- Owned classifier factories reuse the adapter for their hard/supervised terms (KD, feature-KD, MoE, Mixup, CutMix; Born-Again inherits through KD). Soft KL / auxiliary terms and update ownership unchanged.
- Custom loss modules, `NLLLoss` subclasses with their own `forward`, custom `train_step_fn`s and `lr_finder` callables keep the raw-input contract (documented; `lr_finder` docstring shows the log-softmax recipe).
- Example 01: bounded `native_nll_workflow()` (`tests/test_examples_smoke.py -k native_nll_workflow`) — fixed-logit probe → 0.31326166, one training epoch, BEST reload, explicit `F.nll_loss(F.log_softmax(raw))` reference. Docs: `concepts.md` §3 + KD paragraph, `quickstart.md` §2.10, examples catalog, CHANGELOG, API reference.

**PRs:** #275 (develop, merged), #276 (main, merged).

**Checks run (both branches):**
- `uv run --frozen pytest -q` → 1618 passed, 1 skipped, 2 deselected (the two libcairo-dependent `test_docs_projection` tests — pre-existing local-env failures, unrelated; CI runs them and passed). New: `test_nll_normalizes_builtin_logits`, `test_nll_preserves_custom_subclass_input`, `test_nll_default_step_matches_cross_entropy` (incl. [2, 2, 1] accumulation microbatches, direct `TrainStepContext`, BEST reload), `test_n7_nll_matches_log_softmax_reference[mean|sum]` (weights [1, 7] + ignore_index), `test_n7_nll_transformer_token_logits_use_class_axis`, `test_kd_hard_term_uses_normalized_nll[0.0|0.5]`, `test_feature_kd_hard_term_uses_normalized_nll`, `test_moe_supervised_term_uses_normalized_nll`, `test_augmentation_supervised_terms_use_normalized_nll[mixup|cutmix]`.
- `ruff check` → All checks passed; `ruff format --check` → all formatted; `pyright --warnings` → 0 errors, 0 warnings
- docs chain (`build_api_reference` / `build_docs` / `mkdocs build --strict` / `build_wiki --check` / projection tests) → pass
- CI: lint-and-test on Python 3.10–3.14, pip-audit, CodeQL, docs build all green on #275 and #276.

**Not verifiable locally:** the CUDA AMP branch of `default_train_step` carries the same one-line adapter call as the CPU branch (under autocast `log_softmax` runs in FP32), but no CUDA device was available here, so that branch is covered by inspection and the shared unit path only. Live Pages/wiki publication is left to the `docs.yml` main-push job.
